### ISSUE TITLE
Nullable ignoring explicit null checks when variable is considered nullable.

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#32252]
```C#
#nullable enable
public class C
{
	void M(object o)
	{
		if (o == null) o.ToString();
	}
}
```

Currently, no warning is reported for this code. This is especially bad for code with W warnings turned off, such that variable declarations not having annotations is considered normal.

/cc @MadsTorgersen, do you agree that we should have a safety warning here? This is related to the reachability topic I added to the agenda.

**Comment by MadsTorgersen** [comment]
I think I agree. It does mean that merely _asking_ about the nullness of a nonnullable variable makes it "maybe null", which may seem odd. But it makes sense too; why did you ask if you weren't worried that it might be true?

I think this is one to discuss in conjunction with your reachability agenda item.

**Comment by 333fred** [comment]
To me that makes sense: if the code makes it to this path, then it _is_ null, regardless of how it was declared. We should warn accordingly.

**Comment by gafter** [comment]
I thought we did this for a while after the LDM decided "because you said so" and found it had unfortunate consequences:

``` c#
void M(string s)
{
    var length1 = s?.Length; // unnecessary null test here makes s nullable afterwards!
    var length2 = s.Length; // warning: possible deference of null
}
```

But I see that last month the LDM looked at this scenario and appeared to confirm this; see  https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-12-05.md in the third-from-last quoted code block:

``` c#
void M(string p1)
{
    if (p1 == null)
    {
        string x = p1; // W warning? -> Yes
    }
}
```

I don't know how to reconcile these.

**Comment by alrz** [comment]
> ```cs
> void M(string s)
> {
>     var length1 = s?.Length; // unnecessary null test here makes s nullable afterwards!
>     var length2 = s.Length; // warning: possible deference of null
> }
> ```

I think this would be a desirable behavior. Imagine the two lines are not that close, so you actually checked for null in one place and forget it on the other. There's two possible fix here:
- (1) IDE can help here by suggesting the type `string?` for `s`, 
- (2) or it can be fixed by removing the redundant null check (with the non-nullable `s` in place)

**Comment by jcouv** [comment]
Closing this issue as duplicate of https://github.com/dotnet/roslyn/issues/33526 (Pure null tests should affect both branches)
